File: 序号/994.腐烂的橘子.cpp
class Solution {
  int dirs[4][2]{{0, 1}, {0, -1}, {1, 0}, {-1, 0}};

 public:
  /** 0空 1good 2 bad 返回全bad的分钟数 */
  int orangesRotting(vector<vector<int>>& grid) {
    // 多源bfs，先讲烂橘子加进队列，每次向四方向扩散，将四方向的新鲜橘子加进队列。
    // 队列为空，还有新鲜橘子，返回-1 ； 最后一个新鲜橘子被传到，返回step
    queue<pair<int, int>> q;
    int rows = grid.size(), cols = grid[0].size(), goodCnt = 0, x, y, nx, ny,
        step = -1;
    for (int i = 0; i < rows; i++) {
      for (int j = 0; j < cols; j++) {
        if (grid[i][j] == 2) {
          q.emplace(i, j);
        } else if (grid[i][j] == 1) {
          goodCnt++;
        }
      }
    }
    if(goodCnt==0) return 0; // 没有好的，直接返回时间0
    if (q.empty()) return -1;// 没有坏的，返回-1，不可能；
    while (!q.empty()) {
      size_t len = q.size();
      step++;
      for (int i = 0; i < len; i++) {
        x = q.front().first;
        y = q.front().second;
        q.pop();
        if (grid[x][y] == 1) {
          grid[x][y] = -1;
          goodCnt--;
          if(goodCnt==0) return step;
        }
        for (auto& dir : dirs) {
          nx = x + dir[0];
          ny = y + dir[1];
          if (nx >= 0 && nx < rows && ny >= 0 && ny < cols && grid[nx][ny] == 1)
            q.emplace(nx, ny);
        }
      }
    }
    return -1;
  }
};

File: 序号/994. 腐烂的橘子.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        total = 0
      
        q = deque()
        vis=set()
        m,n=len(grid),len(grid[0])
        # 找出所有橘子和腐烂橘子的坐标再bfs
        for i,row in enumerate(grid):
            for j,x in enumerate(row):
                if x>0:total+=1
                if x==2:
                    q.append((i,j))
                    vis.add((i,j))
        lj = len(q)
        dirs=(-1,0,1,0,-1)
        minute = 0
        while q:
            sz = len(q)
            
            for _ in range(sz):
                r,c =q.popleft()
                for d in range(4):
                    nr=r+dirs[d]
                    nc=c+dirs[d+1]
                    if 0<=nr<m and 0<=nc<n:
                        if (nr,nc) in vis:continue
                        if grid[nr][nc]==1:
                            vis.add((nr,nc))
                            q.append((nr,nc))
                            lj+=1
            if q:minute+=1
        return minute if lj==total else -1